### PR TITLE
Use a shared `netstat` snapshot for DNS socket checks to avoid races and redundant calls

### PR DESCRIPTION
**Pull request by jumpsmm7**
### Motivation
- Reduce races and repeated `netstat` calls when inspecting port 53 ownership by collecting a single snapshot and sharing it across DNS ownership checks.
- Avoid writing predictable temporary files for the snapshot and preserve trailing newlines from `netstat` output for reliable parsing.

### Description
- Add `dns_socket_snapshot` and `dns_socket_snapshot_value` helpers to capture a single `netstat -nlp` snapshot into `DNS_SOCKET_SNAPSHOT` and mark its validity.
- Update DNS ownership and readiness helpers (`adguardhome_owns_dns`, `kill_dns_port_owners`, `dns_port_available`, `dns_port_has_foreign_owner`, `dns_port_needs_release`, `release_dns_port_from_dnsmasq`, `log_adguardhome_dns_wait_failure`, `wait_for_adguardhome_dns`, and `start_dns_port_guard`) to accept and use the shared snapshot when provided and to refresh it when required.
- Adjust error/return handling to rely on snapshot validity instead of implicitly recollecting output or using `adguardhome_single_process_running` in fallback paths.
- Update `S99AdGuardHome.md5sum` and `S99AdGuardHome.sha256sum` to reflect the modified script and extend tests to record `netstat` calls.

### Testing
- Ran `tests/s99-netstat-readiness.sh` which exercises shared snapshot collection and readiness checks and it passed.
- Ran `tests/dns-startup-handoff.sh` which verifies handoff ordering, snapshot refresh after kills, and netstat call counts and it passed.
- Ran `tests/s99-dns-mode-lifecycle.sh` which simulates lifecycle behavior using the shared snapshot and it passed.

------
[Codex Task](https://chatgpt.com/codex/cloud/tasks/task_e_6a622a3b27a88329b0e8e4f2a5eb864f)

### CODE REVIEW

**Comment by jumpsmm7** [comment]
@coderabbitai approve

**Comment by jumpsmm7** [comment]
@codex address all feedback, update checksums, and all regression test pathways

**Comment by jumpsmm7** [comment]
@codex total review

**Comment by jumpsmm7** [comment]
@codex address all feedback, update checksums, and all regression test pathways

**Comment by jumpsmm7** [comment]
@codex total review

**Comment by jumpsmm7** [comment]
@coderabbitai full review

**Comment by jumpsmm7** [comment]
@coderabbitai full review

**Comment by jumpsmm7** [comment]
@codex review focus on edge cases, performance optimizations, security issues, security vulnerabilities, security concerns, behavioral regressions, missing tests, improper code behaviors, CI issues

File: S99AdGuardHome [hunk 1]
```diff
@@ -1506,15 +1580,38 @@ pre_start_adguardhome() {
 	_dns_watchdog_saved_traps="${DNS_WATCHDOG_TRAP_FILE}"
 	agh_log INFO pre_start_adguardhome "Stopping dnsmasq/releasing port 53 before starting AdGuardHome."
 	while [ "${_dns_stop_attempts}" -lt "${_dns_stop_limit}" ]; do
-		if dns_port_available "${_dns_bind_scope}"; then
+		_dns_snapshot_attempts="0"
+		while ! dns_socket_snapshot; do
```
[the remark this answers is not in the document]
Comment (jumpsmm7) [reply S99AdGuardHome]: @codex resolve this P2

File: S99AdGuardHome [hunk 2]
```diff
@@ -985,20 +1018,24 @@ dns_port_has_foreign_owner() {
...
 		$0 ~ /^(tcp|udp)6?[[:space:]]+/ && $0 ~ /:53[[:space:]]/ && scoped($4) {
+			bound = 1
+			owner = ""
 			for (field = 1; field <= NF; field++) {
 				if ($field ~ /^[0-9]+\/[^[:space:]]+$/) {
+					owner = $field
 					if ($field !~ "/" proc "$") foreign_owner = 1
 					break
 				}
 			}
```
[the remark this answers is not in the document]
Comment (jumpsmm7) [reply S99AdGuardHome:1030]: @codex resolve this P2, update checksums, and all regression test pathways.

**Comment by jumpsmm7** [comment]
@codex review for CI bugs and vulnerabilities

**Comment by jumpsmm7** [comment]
@coderabbitai unpause

**Comment by jumpsmm7** [comment]
@coderabbitai full review

**Comment by jumpsmm7** [comment]
@coderabbitai full review

**Comment by jumpsmm7** [comment]
@codex fix Code Quality / Shell and checksum quality checks

**Comment by jumpsmm7** [comment]
@codex review for bugs in Code Quality / Shell and checksum quality checks

**Comment by jumpsmm7** [comment]
@codex fix I am the GitHub Copilot Chat Assistant.

Summary of root cause
- The failing job (ID 89450312290, workflow .github/workflows/code-quality.yml) fails because the test "AdGuardHome DNS startup handoff regression" reports:
  FAIL: pre-start did not release the DNS owner after retrying snapshot collection
- That means the pre-start logic that performs snapshot collection / retries is not cleaning up the DNS ownership/lock/port after a failed attempt. A subsequent attempt (or the test harness) expects the pre-start to release ownership so the next step can bind DNS, but it remains held, causing the handoff failure and the test to fail.

Concrete fixes (what to change)
1. Make release/cleanup of the DNS owner always run after any failed snapshot-collection attempt.
   - Ensure the cleanup is idempotent (safe to call multiple times) and tolerant of missing files/processes.
   - Use a trap on EXIT/INT/TERM in the pre-start script so cleanup always runs on exit.
   - In the retry loop, explicitly call the cleanup routine before the next attempt/sleep.

2. Ensure the release routine actually frees any mechanism used to "own" DNS:
   - If ownership is represented by a pidfile, remove it.
   - If by a process binding the port, kill or stop that process (use fuser/lsof to identify owner) or wait for the socket to be closed before continuing.
   - If a lock (flock) is used, ensure the fd is closed or the lock file removed.

3. Strengthen the test to assert cleanup and add a short wait with timeouts when checking for port release, to avoid flaky timing races.

Suggested code changes
- Add an idempotent cleanup function and trap.
- Call cleanup inside the retry loop before the next attempt.

Example POSIX-sh compatible patch (insert in the pre-start / snapshot collection script; rename variables/paths to match your code):

Add at top of script (or before any early exits):

cleanup_release_dns_owner() {
  # idempotent cleanup: remove pidfile/lock and try to free the port owner
  : "${DNS_OWNER_FILE:=/var/run/adguard_dns_owner.pid}"
  : "${DNS_PORT:=53}"
  # remove pidfile/lock if present
  if [ -f "$DNS_OWNER_FILE" ]; then
    rm -f "$DNS_OWNER_FILE" || true
  fi

  # if fuser is available, try to kill any process still holding the DNS UDP port
  if command -v fuser >/dev/null 2>&1; then
    # fuser will return non-zero if no process found; ignore errors
    fuser -k "${DNS_PORT}/udp" 2>/dev/null || true
    fuser -k "${DNS_PORT}/tcp" 2>/dev/null || true
  elif command -v lsof >/dev/null 2>&1; then
    # fallback with lsof
    lsof -iUDP:"$DNS_PORT" -t 2>/dev/null | xargs -r kill -9 2>/dev/null || true
    lsof -iTCP:"$DNS_PORT" -t 2>/dev/null | xargs -r kill -9 2>/dev/null || true
  fi

  # wait for the port to be free (short timeout)
  timeout=5
  while [ "$timeout" -gt 0 ]; do
    # check no process is bound to the port
    if ! (command -v ss >/dev/null 2>&1 && ss -lun | grep -q ":$DNS_PORT ") && \
       ! (command -v netstat >/dev/null 2>&1 && netstat -lun 2>/dev/null | grep -q ":$DNS_PORT "); then
      break
    fi
    timeout=$((timeout - 1))
    sleep 1
  done
}

# ensure cleanup always runs on exit/interruption
trap 'cleanup_release_dns_owner' EXIT INT TERM

Then change the retry loop that attempts snapshot collection to call cleanup before next attempt:

max_attempts=5
attempt=1
while ! snapshot_collect; do
  echo "snapshot_collect failed (attempt $attempt)"
  # release any DNS owner/locks before retrying
  cleanup_release_dns_owner

  attempt=$((attempt + 1))
  if [ "$attempt" -gt "$max_attempts" ]; then
    echo "FAIL: pre-start did not release the DNS owner after retrying snapshot collection" >&2
    exit 1
  fi
  sleep 1
done

Notes and rationale
- The trap ensures cleanup runs even if the script exits unexpectedly.
- Calling cleanup between retries ensures successive attempts get a clean environment (the current failing test expects that).
- The kill attempts and port checks are defensive: if the test harness or earlier steps left a listener on the DNS port, they will be removed and the port freed.
- Make cleanup idempotent to avoid errors when it is run multiple times or when files/processes are already gone.

Test adjustments
- Modify the failing test to assert the existence of cleanup: after a failed snapshot_collect attempt the test should check that:
  - The pidfile/lock does not exist, and
  - The DNS port is not bound before proceeding (or wait up to N seconds for it to be freed).
- Add diagnostics on failure (e.g., dump of lsof/ss/netstat output and the pidfile contents) to make future runs easier to debug.

Where to apply this
- Locate the pre-start / snapshot-collection script(s) in your repo (grep for "snapshot collection", "pre-start", "DNS owner", or where the code prints "snapshot collection" or handles DNS handoff). Update that file with the pattern shown above.
- If you use a helper library for installer lifecycle tasks, add the cleanup function to that library and call it from pre-start and from tests’ teardown.

How to validate locally
- Run the test suite locally (the CI runs code-quality, but the failing test is from your internal test suite). If you have a script to run regressions, run the specific "AdGuardHome DNS startup handoff regression" test after applying the patch.
- Re-run the full CI workflow (push or open a PR) to confirm the job succeeds.

If you want, I can produce a precise diff for the specific file once you point me to the script path that implements pre-start/snapshot collection (or give me the filenames to patch).

**Comment by jumpsmm7** [comment]
@coderabbitai full review

**Comment by jumpsmm7** [comment]
@codex review
